File: Python/_13_Roman_to_Integer.py
class Solution:
    """
    A solution that converts a Roman numeral string into an integer.
    This version explicitly checks each numeral and its subtractive cases 
    (like IV = 4, IX = 9, etc.) using conditional statements.
    """

    def romanToInt(self, s: str) -> int:
        """
        Convert a Roman numeral string to an integer.

        Args:
            s (str): Roman numeral string (e.g., "MCMXCIV")

        Returns:
            int: The integer value of the Roman numeral
        """
        result = 0  # Final accumulated integer value
        n = len(s)
        skip = False  # Used to skip the next character if we've processed a subtractive pair

        for i in range(n):
            # Skip if previous iteration already consumed this numeral in a pair
            if skip:
                skip = False
                continue

            # Direct matches and subtractive cases
            if s[i] == "M":
                result += 1000

            if s[i] == "C" and i < n - 1 and s[i + 1] == "M":
                result += 900
                skip = True
                continue

            if s[i] == "D":
                result += 500

            if s[i] == "C" and i < n - 1 and s[i + 1] == "D":
                result += 400
                skip = True
                continue

            if s[i] == "C":
                result += 100

            if s[i] == "X" and i < n - 1 and s[i + 1] == "C":
                result += 90
                skip = True
                continue

            if s[i] == "L":
                result += 50

            if s[i] == "X" and i < n - 1 and s[i + 1] == "L":
                result += 40
                skip = True
                continue

            if s[i] == "X":
                result += 10

            if s[i] == "I" and i < n - 1 and s[i + 1] == "X":
                result += 9
                skip = True
                continue

            if s[i] == "V":
                result += 5

            if s[i] == "I" and i < n - 1 and s[i + 1] == "V":
                result += 4
                skip = True
                continue

            if s[i] == "I":
                result += 1

        return result


class BetterSolution:
    """
    A more elegant and concise solution that uses a dictionary lookup
    and a subtraction rule: if a smaller numeral appears before a larger one, subtract it.
    """

    def romanToInt(self, s: str) -> int:
        """
        Convert a Roman numeral string to an integer.

        Args:
            s (str): Roman numeral string (e.g., "MCMXCIV")

        Returns:
            int: The integer value of the Roman numeral
        """
        result = 0

        # Map of Roman numerals to integer values
        nums = {"I": 1, "V": 5, "X": 10, "L": 50,
                "C": 100, "D": 500, "M": 1000}

        for i in range(len(s)):
            # If the current numeral is smaller than the next one, subtract it
            if i + 1 < len(s) and nums[s[i]] < nums[s[i + 1]]:
                result -= nums[s[i]]
            else:
                result += nums[s[i]]

        return result


def main():
    """
    Test both solutions with the Roman numeral "MCMXCIV" (1994).
    """
    sol = Solution()
    sol1 = BetterSolution()

    # Using the explicit, verbose solution
    print(sol.romanToInt("MCMXCIV"))  # Expected output: 1994

    # Using the concise dictionary-based solution
    print(sol1.romanToInt("MCMXCIV"))  # Expected output: 1994


if __name__ == "__main__":
    main()

File: C#/_13_Roman_to_Integer.cs
using System;

// _13_Roman_to_Integer.Solution sol = new _13_Roman_to_Integer.Solution();
// _13_Roman_to_Integer.AnotherSolution aSol = new _13_Roman_to_Integer.AnotherSolution();
// Console.WriteLine(sol.romanToInt("MMMXLVIII"));
// Console.WriteLine(aSol.romanToInt("MMMXLVIII"));

/// <summary>
/// Problem 13: Roman to Integer
/// Contains two different implementations for converting a Roman numeral string into its integer value.
/// </summary>
public class _13_Roman_to_Integer
{
    /// <summary>
    /// First approach using a Dictionary for Roman numeral values.
    /// </summary>
    public class Solution
    {
        /// <summary>
        /// Converts a Roman numeral string to an integer.
        /// </summary>
        /// <param name="s">Roman numeral string (e.g., "XIV")</param>
        /// <returns>Integer value of the Roman numeral</returns>
        public int romanToInt(string s)
        {
            int result = 0;

            // Dictionary mapping Roman numeral characters to their integer values
            Dictionary<char, int> nums = new Dictionary<char, int>();
            nums.Add('I', 1); nums.Add('V', 5); nums.Add('X', 10); nums.Add('L', 50);
            nums.Add('C', 100); nums.Add('D', 500); nums.Add('M', 1000);

            // Loop through each character in the string
            for (int i = 0; i < s.Length; i++)
            {
                int num1; // Current numeral value
                int num2 = 0; // Next numeral value (default 0 if none)

                // Look up the value for the current character
                nums.TryGetValue(s[i], out num1);

                // If not at the end, also get the next character’s value
                if (i + 1 < s.Length)
                {
                    nums.TryGetValue(s[i + 1], out num2);
                }

                // If current is smaller than next, subtract (e.g., IV = 5 - 1 = 4)
                if (i + 1 < s.Length && num1 < num2)
                {
                    result -= num1;
                }
                else // Otherwise, just add it
                {
                    result += num1;
                }
            }

            return result;
        }
    }

    /// <summary>
    /// Second approach using an integer array indexed by character codes
    /// for faster lookup (avoids Dictionary overhead).
    /// </summary>
    public class AnotherSolution
    {
        // Precomputed values for ASCII character codes
        private static readonly int[] nums;

        // Static constructor initializes the lookup table
        static AnotherSolution()
        {
            nums = new int[89]; // Large enough to cover ASCII up to 'X'/'M'
            nums['I'] = 1; nums['V'] = 5; nums['X'] = 10; nums['L'] = 50;
            nums['C'] = 100; nums['D'] = 500; nums['M'] = 1000;
        }

        /// <summary>
        /// Converts a Roman numeral string to an integer.
        /// </summary>
        /// <param name="s">Roman numeral string (e.g., "MCMXCIV")</param>
        /// <returns>Integer value of the Roman numeral</returns>
        public int romanToInt(string s)
        {
            int result = 0;

            // Loop through each character
            for (int i = 0; i < s.Length; i++)
            {
                // If the current numeral is less than the next, subtract it
                if (i + 1 < s.Length && nums[s[i]] < nums[s[i + 1]])
                {
                    result -= nums[s[i]];
                }
                else // Otherwise, add it
                {
                    result += nums[s[i]];
                }
            }

            return result;
        }
    }
}